Add async function support

scothis opened the pull request:
Either async or Promised function can now be used in addition to
traditional synchronous functions. Node-style callback functions are
not supported.

```js
// sync
module.exports = name => `Hello ${name}!`;

// promise
module.exports = name => Promise.resolve(`Hello ${name}!`);

// async
module.exports = async name => `Hello ${name}!`;
```

Issue: #7

jldec commented:
@scothis thank you - this looks like it solves #7.

Do you have any thoughts on what the recommended approach would be to "promisify" or "asyncify" a node-style callback? I'd like to include an example of a function which uses one of the node async callback apis (like fs.readFile), preferably without relying on any 3rd party libraries.

scothis commented:
@jldec yea, callbacks are possible but not nearly as elegant. As you mentioned there are third party libraries that help with this, all they really do is apply this pattern in a more generic form. Without a library bridging callbacks to Promises looks like:

```js
const fs = require('fs');

module.exports = path => new Promise((resolve, reject) => {
    fs.readFile(path, (err, data) => {
        if (err) return reject(err);
        resolve(data);
    });
});
```

A more generic form:

```js
function promisify(fn) {
    return value => new Promise((resolve, reject) => {
        fn(value, (err, result) => {
            if (err) return reject(err);
            resolve(result);
        });
    });
}
```

Ultimately, I think proper callback support will be necessary, but it will require a configuration option so that the function invoker knows how to invoke the function. Trying to support both patterns without a config switch creates a footgun.

jldec commented:
@scothis - fyi - there is also node8 [util.promisify](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original)

jldec commented:
the async support checks out (tested in riff) but i'm getting a weird unit test error.
could you validate if you get this also on your end @scothis ?

```
$ npm test

> node-function-invoker@1.0.0 test ~/riff/node-function-invoker/src/main/node
> jasmine

Started
Error
    at makeApp (~/riff/node-function-invoker/src/main/node/spec/appSpec.js:9:41)
    at app.post.asyncMiddleware (~/riff/node-function-invoker/src/main/node/lib/app.js:23:31)
    at ~/riff/node-function-invoker/src/main/node/lib/app.js:10:9
    at Layer.handle [as handle_request] (~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/layer.js:95:5)
    at next (~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/layer.js:95:5)
    at ~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/index.js:335:12)
    at next (~/riff/node-function-invoker/src/main/node/node_modules/express/lib/router/index.js:275:10)
.Result Hello riff!
.Result Hello riff!
.Result Hello riff!
.Result Hello riff!
.Result [object Object]
.Result [object Object]
.


7 specs, 0 failures
Finished in 0.07 seconds
```

scothis commented:
The stack trace is coming from the [test confirming that errors are handled](https://github.com/projectriff/node-function-invoker/pull/9/files#diff-8388b122e673d29d56b5c5975c5ad00cR9). It's not a test failure.

I'll take a look later today to see if the stack trace can be suppressed from the output.

jldec commented:
Oh, good, i noticed it wasn't actually failing but the error messages still looked suspicious - would be good to suppress those.

scothis commented:
@jldec I added a commit that will suppress all debug logging by default. We should think more about how real errors are exposed. This isn't ideal for production as the error is swallowed.

The stack trace was coming from express' default error handler. I added a custom error handler to log the error if debug logging is enabled and send a 500 response.

jldec commented:
Awesome, i really like the addition of the 500 handler (instead of exiting node).

I also like the use of a util.debuglog logger to control verbosity for things like the logging of function results. (i would have used stdout instead of stderr, but let's stick to node's convention here)

Suppressing all errors unless the environment is configured not to, feels a bit extreme. How about replacing the util.debuglog logic (designed for debugging?) with a conditional call to console.error(). e.g.

```js
    // handle errors
    app.use((err, req, res, next) => {
        if (!process.env.SUPPRESS_ERRORS) { console.error('Error:', err); }
        res.status(500).end();
    });
```

scothis commented:
@jldec Let's take the discussion around logging into a new issue/PR. We should establish/follow the logging patterns for all function invokers.

I agree that we want to see unhandled errors in production, but I'm hesitant to add more configuration unless the pattern is going to be adopted across all function invokers.

markfisher commented:
This issue is also relevant to the extent that we need to address error handling consistently across all languages so that things bubble up in a consistent way as well: https://github.com/projectriff/http-gateway/issues/22

jldec commented:
@scothis - i'll take your point about separating the error logging strategy. PR merged! :)
